daffodil-core/src/main/scala/edu/illinois/ncsa/daffodil/dsom/AnnotatedSchemaComponent.scala

```diff
@@ -41,23 +41,122 @@ import edu.illinois.ncsa.daffodil.equality._
+        // supply constructor's last arg is boolean indicating it's a default property
+        Found(value, loc, pname, true)
+      case NotFound(nd, d, pn) =>
+        Assert.invariant(d.isEmpty)
+        NotFound(Seq(), nd, pn) // we want the places we searched shown as default locations searched
+    }
+    fixup
+  }
+
+  override def findPropertyOption(pname: String): PropertyLookupResult = {
+    ExecutionMode.requireCompilerMode
+    // first try in regular properties
+    val regularResult = resolver.findNonDefaultProperty(pname)
+    regularResult match {
+      case f: Found => f
+      case NotFound(nonDefaultLocsTried1, defaultLocsTried1, _) => {
+        Assert.invariant(defaultLocsTried1.isEmpty)
+        val defaultResult = resolver.findDefaultProperty(pname)
+        defaultResult match {
+          case f: Found => f
+          case NotFound(nonDefaultLocsTried2, defaultLocsTried2, _) => {
+            Assert.invariant(nonDefaultLocsTried2.isEmpty)
+            // did not find it at all. Return a NotFound with all the places we
+            // looked non-default and default.
+            val nonDefaultPlaces = nonDefaultLocsTried1
+            val defaultPlaces = defaultLocsTried2
+            NotFound(nonDefaultPlaces, defaultPlaces, pname)
+          }
+        }
+      }
+    }
+  }
+}
+
+abstract class AnnotatedSchemaComponentImpl( final override val xml: Node,
+  final override val parent: SchemaComponent)
+  extends AnnotatedSchemaComponent
 
 /**
  * Shared characteristics of any annotated schema component.
  * Not all components can carry DFDL annotations.
  */
+trait AnnotatedSchemaComponent
+  extends SchemaComponent
+  with AnnotatedMixin
+  with OverlapCheckMixin {
 
-abstract class AnnotatedSchemaComponent(xml: Node, sc: SchemaComponent)
-  extends SchemaComponent(xml, sc)
-  with AnnotatedMixin {
+  final lazy val term: Term = this match {
+    case gr: GroupRef => gr.asModelGroup
+    //    case mg: ModelGroup => mg.parent match {
+    //      case ggd: GlobalGroupDef => {
+    //        // this is the model group that is the definition of
+    //        // a global group, so our term is the group reference referring to this.
+    //        ggd.groupRef.asModelGroup
+    //      }
+    //      case _ => mg
+    //    }
+    case t: Term => t
+    case ged: GlobalElementDecl => ged.elementRef
+    case ty: SimpleTypeDefBase => ty.elementBase
+    case ty: ComplexTypeBase => ty.elementBase
+    case ggd: GlobalGroupDef => ggd.groupRef.asModelGroup
+    case sd: SchemaDocument =>
+      Assert.usageError("not to be called for schema documents")
+  }
+
+  final lazy val resolver: ResolvesProperties = {
+    val res = this match {
+      case sd: SchemaDocument => sd
+      case _ => term
+    }
+    res
+  }
 
   requiredEvaluations(annotationObjs)
   requiredEvaluations(shortFormPropertiesCorrect)
   requiredEvaluations(nonDefaultPropertySources)
   requiredEvaluations(defaultPropertySources)
 
-  def term: Term
```

Comment:
Add scaladoc. What does term do if the annotated schema component is say, a variable declaration?

Reply:
Actually , not well formed question - a variable declarations are not Annotated Schema Components, because they themselves are DFDL annotations.